File: clients/python/src/virtualfs/_http.py
"""Low-level HTTP transport.

Wraps `httpx.Client` with:
  - Bearer token auth (lazily bootstrapped from `auth_secret` if needed)
  - Retry on transient 5xx / 429 with exponential backoff
  - Mapping of HTTP status codes to typed `VirtualFSError` subclasses

Higher layers (Client, Sandbox, FilesAPI) should never touch httpx directly —
they always go through `Transport.request()` / `.stream()`.
"""

from __future__ import annotations

import json
import random
import time
from collections.abc import Iterator, Mapping
from typing import Any, Dict, Optional, Union
from urllib.parse import quote

import httpx

from ._version import __version__
from .errors import (
    AuthError,
    ConflictError,
    ExecTimeoutError,
    NotFoundError,
    RateLimitError,
    ServerError,
    TransportError,
    ValidationError,
    VirtualFSError,
)

# Default retry policy. Idempotent + transient-failure paths only.
DEFAULT_MAX_RETRIES = 3
RETRY_STATUS = frozenset({429, 500, 502, 503, 504})
DEFAULT_TIMEOUT_S = 30.0


JsonType = Union[Dict[str, Any], list, str, int, float, bool, None]


def encode_path(path: str) -> str:
    """Encode a sandbox-relative path for use in `/files/{path}`.

    The server expects no leading slash and treats `/` as a separator, so we
    only quote characters that would break URL parsing — slashes are safe.
    """
    return quote(path.lstrip("/"), safe="/")


class Transport:
    """Shared HTTP plumbing for `Client` and `Sandbox`."""

    def __init__(
        self,
        *,
        base_url: str,
        token: Optional[str] = None,
        auth_secret: Optional[str] = None,
        admin_secret: Optional[str] = None,
        sub: Optional[str] = None,
        tenant: Optional[str] = None,
        expires_in: str = "30d",
        timeout: float = DEFAULT_TIMEOUT_S,
        max_retries: int = DEFAULT_MAX_RETRIES,
        user_agent: Optional[str] = None,
        http_client: Optional[httpx.Client] = None,
    ) -> None:
        if not (token or auth_secret or admin_secret):
            raise ValueError(
                "Provide one of: token=..., auth_secret=..., or admin_secret=..."
            )
        if (auth_secret or admin_secret) and not sub:
            raise ValueError(
                "`sub` is required when bootstrapping a token from a secret"
            )

        self._base_url = base_url.rstrip("/")
        self._token = token
        self._auth_secret = auth_secret
        self._admin_secret = admin_secret
        self._sub = sub
        self._tenant = tenant
        self._expires_in = expires_in
        self._max_retries = max_retries
        self._owns_client = http_client is None
        self._http = http_client or httpx.Client(timeout=timeout)
        self._user_agent = user_agent or f"virtualfs-python/{__version__}"

    # ── lifecycle ────────────────────────────────────────────────────────────
    def close(self) -> None:
        if self._owns_client:
            self._http.close()

    def __enter__(self) -> Transport:
        return self

    def __exit__(self, *_exc: Any) -> None:
        self.close()

    # ── auth ─────────────────────────────────────────────────────────────────
    @property
    def token(self) -> str:
        if self._token is None:
            self._token = self._bootstrap_token()
        return self._token

    def _bootstrap_token(self) -> str:
        """Exchange `AUTH_SECRET` (or `ADMIN_SECRET`) for a JWT.

        Uses `POST /v1/auth/bootstrap` with `X-Auth-Secret` for the standard
        path; falls back to `POST /v1/auth/admin` with `X-Admin-Secret` when
        only `admin_secret` is provided.
        """
        body: Dict[str, Any] = {"sub": self._sub, "expiresIn": self._expires_in}

        if self._auth_secret is not None:
            if self._tenant:
                body["tenant"] = self._tenant
            url = f"{self._base_url}/v1/auth/bootstrap"
            headers = {"X-Auth-Secret": self._auth_secret}
        else:
            assert self._admin_secret is not None
            url = f"{self._base_url}/v1/auth/admin"
            headers = {"X-Admin-Secret": self._admin_secret}

        resp = self._http.post(
            url,
            json=body,
            headers={**headers, "User-Agent": self._user_agent},
        )
        if resp.status_code != 201:
            self._raise_for_status(resp)
        data = resp.json()
        token = data.get("token")
        if not isinstance(token, str):
            raise AuthError(
                "bootstrap response missing token", status=resp.status_code, details=data
            )
        return token

    def _auth_headers(self) -> Dict[str, str]:
        return {
            "Authorization": f"Bearer {self.token}",
            "User-Agent": self._user_agent,
        }

    # ── request execution ────────────────────────────────────────────────────
    def request(
        self,
        method: str,
        path: str,
        *,
        json_body: Optional[JsonType] = None,
        content: Optional[bytes] = None,
        params: Optional[Mapping[str, Any]] = None,
        headers: Optional[Mapping[str, str]] = None,
        files: Optional[Mapping[str, Any]] = None,
        timeout: Optional[float] = None,
        expect_json: bool = True,
    ) -> httpx.Response:
        """Send a request with retry + error mapping.

        `path` is appended to `base_url + /v1`. Pass either `json_body` (sets
        Content-Type to application/json) or `content` (raw bytes — caller
        supplies Content-Type via `headers`).
        """
        url = f"{self._base_url}/v1{path}"
        merged_headers: Dict[str, str] = self._auth_headers()
        if headers:
            merged_headers.update(headers)

        last_exc: Optional[Exception] = None
        for attempt in range(self._max_retries + 1):
            try:
                resp = self._http.request(
                    method,
                    url,
                    json=json_body if content is None else None,
                    content=content,
                    params=params,
                    headers=merged_headers,
                    files=files,
                    timeout=timeout,
                )
            except httpx.TransportError as e:
                last_exc = e
                if attempt >= self._max_retries:
                    raise TransportError(
                        f"network error after {attempt + 1} attempts: {e}"
                    ) from e
                self._sleep_backoff(attempt)
                continue

            if resp.status_code in RETRY_STATUS and attempt < self._max_retries:
                # Honour Retry-After when present, else exponential jitter.
                retry_after = _parse_retry_after(resp)
                if retry_after is not None:
                    time.sleep(min(retry_after, 30))
                else:
                    self._sleep_backoff(attempt)
                continue

            if resp.status_code >= 400:
                self._raise_for_status(resp)
            return resp

        # Unreachable — loop either returns or raises.
        raise TransportError(f"unexpected retry exhaustion: {last_exc}")

    def stream(
        self,
        method: str,
        path: str,
        *,
        json_body: Optional[JsonType] = None,
        params: Optional[Mapping[str, Any]] = None,
        headers: Optional[Mapping[str, str]] = None,
        timeout: Optional[float] = None,
    ) -> httpx.Response:
        """Open a streaming response (used for SSE exec and tar.gz export).

        The caller is responsible for `.close()` on the response when done.
        Streaming responses are NOT retried — at-most-once semantics.
        """
        url = f"{self._base_url}/v1{path}"
        merged_headers = self._auth_headers()
        if headers:
            merged_headers.update(headers)

        req = self._http.build_request(
            method,
            url,
            json=json_body,
            params=params,
            headers=merged_headers,
            timeout=timeout,
        )
        resp = self._http.send(req, stream=True)
        if resp.status_code >= 400:
            # Body must be drained before mapping the error; tiny payloads.
            try:
                resp.read()
            finally:
                resp.close()
            self._raise_for_status(resp)
        return resp

    # ── error mapping ────────────────────────────────────────────────────────
    def _raise_for_status(self, resp: httpx.Response) -> None:
        status = resp.status_code
        body, code, message, details = _parse_error_body(resp)
        if status in (401, 403):
            raise AuthError(message, code=code, status=status, details=details)
        if status == 404:
            raise NotFoundError(message, code=code, status=status, details=details)
        if status == 408:
            raise ExecTimeoutError(
                message,
                code=code,
                status=status,
                details=details,
                duration_ms=body.get("durationMs") if isinstance(body, dict) else None,
            )
        if status == 409:
            raise ConflictError(message, code=code, status=status, details=details)
        if status == 400:
            raise ValidationError(message, code=code, status=status, details=details)
        if status == 429:
            raise RateLimitError(
                message,
                code=code,
                status=status,
                details=details,
                retry_after=_parse_retry_after(resp),
            )
        if 500 <= status < 600:
            raise ServerError(message, code=code, status=status, details=details)
        raise VirtualFSError(message, code=code, status=status, details=details)

    # ── retry sleep ──────────────────────────────────────────────────────────
    def _sleep_backoff(self, attempt: int) -> None:
        # Exponential backoff with full jitter: [0, base * 2^attempt)
        base = 0.25
        delay = random.uniform(0, base * (2 ** attempt))
        time.sleep(min(delay, 8.0))


# ── module-level helpers ─────────────────────────────────────────────────────
def _parse_error_body(resp: httpx.Response) -> tuple:
    """Best-effort parse of `{ error, code, details }` JSON response body."""
    try:
        body = resp.json()
    except (json.JSONDecodeError, ValueError):
        body = {}
    if not isinstance(body, dict):
        body = {}
    code = body.get("code")
    message = body.get("error") or f"HTTP {resp.status_code}"
    details = body.get("details")
    return body, code, message, details


def _parse_retry_after(resp: httpx.Response) -> Optional[int]:
    raw = resp.headers.get("Retry-After")
    if raw is None:
        return None
    try:
        return int(raw)
    except ValueError:
        return None


def iter_sse_events(resp: httpx.Response) -> Iterator[tuple]:
    """Iterate SSE events from a streaming response.

    Yields `(event_name, parsed_json)` tuples. `data:` lines are concatenated
    per event and parsed as JSON. Comment lines (`:`) are skipped.
    """
    event = "message"
    data_lines: list = []
    for line in resp.iter_lines():
        if line == "":
            if data_lines:
                raw = "\n".join(data_lines)
                try:
                    payload = json.loads(raw)
                except json.JSONDecodeError:
                    payload = {"raw": raw}
                yield event, payload
            event = "message"
            data_lines = []
            continue
        if line.startswith(":"):
            continue
        if line.startswith("event:"):
            event = line[len("event:"):].strip()
        elif line.startswith("data:"):
            data_lines.append(line[len("data:"):].lstrip(" "))
    # Trailing event without blank line
    if data_lines:
        raw = "\n".join(data_lines)
        try:
            payload = json.loads(raw)
        except json.JSONDecodeError:
            payload = {"raw": raw}
        yield event, payload

File: clients/typescript/src/http.ts
import {
	AuthError,
	ConflictError,
	ExecTimeoutError,
	NotFoundError,
	RateLimitError,
	SQLFSError,
	ServerError,
	TransportError,
	ValidationError,
} from "./errors.js";
import { readJsonObject } from "./internal.js";
import { version } from "./version.js";

export type JsonValue = Record<string, unknown> | unknown[] | string | number | boolean | null;

export interface TransportOptions {
	baseUrl: string;
	token?: string;
	authSecret?: string;
	adminSecret?: string;
	sub?: string;
	tenant?: string;
	expiresIn?: string;
	timeout?: number;
	maxRetries?: number;
	userAgent?: string;
	fetch?: typeof globalThis.fetch;
}

export interface RequestOptions {
	jsonBody?: JsonValue;
	content?: BodyInit;
	params?: Record<string, unknown>;
	headers?: Record<string, string>;
	timeout?: number;
	idempotent?: boolean;
	readOnly?: boolean;
}

const defaultMaxRetries = 3;
const defaultTimeoutMs = 30_000;
const retryStatus = new Set([429, 500, 502, 503, 504]);

export function encodePath(path: string): string {
	return path
		.replace(/^\/+/, "")
		.split("/")
		.map((part) => encodeURIComponent(part))
		.join("/");
}

export class Transport {
	private readonly baseUrl: string;
	private currentToken?: string;
	private tokenPromise?: Promise<string>;
	private readonly authSecret?: string;
	private readonly adminSecret?: string;
	private readonly adminBearerToken?: string;
	private readonly sub?: string;
	private readonly tenant?: string;
	private readonly expiresIn: string;
	private readonly timeout: number;
	private readonly maxRetries: number;
	private readonly userAgent: string;
	private readonly fetchImpl: typeof globalThis.fetch;

	constructor(options: TransportOptions) {
		if (!(options.token || options.authSecret || options.adminSecret)) {
			throw new Error("Provide one of: token, authSecret, or adminSecret");
		}
		if ((options.authSecret || options.adminSecret) && !options.sub) {
			throw new Error("sub is required when bootstrapping a token from a secret");
		}
		if (options.adminSecret && !options.token) {
			throw new Error("token is required with adminSecret because /v1/auth/admin requires Bearer authentication");
		}
		if (options.authSecret && options.adminSecret) {
			throw new Error("authSecret and adminSecret are mutually exclusive");
		}

		this.baseUrl = options.baseUrl.replace(/\/+$/, "");
		this.currentToken = options.adminSecret ? undefined : options.token;
		this.authSecret = options.authSecret;
		this.adminSecret = options.adminSecret;
		this.adminBearerToken = options.adminSecret ? options.token : undefined;
		this.sub = options.sub;
		this.tenant = options.tenant;
		this.expiresIn = options.expiresIn ?? "30d";
		this.timeout = options.timeout ?? defaultTimeoutMs;
		this.maxRetries = options.maxRetries ?? defaultMaxRetries;
		this.userAgent = options.userAgent ?? `sql-fs-sdk/${version}`;
		this.fetchImpl = options.fetch ?? globalThis.fetch;

		if (typeof this.fetchImpl !== "function") {
			throw new Error("No fetch implementation available; pass fetch in Client options");
		}
	}

	get token(): string | undefined {
		return this.currentToken;
	}

	async getToken(): Promise<string> {
		if (this.currentToken !== undefined) {
			return this.currentToken;
		}
		if (this.tokenPromise === undefined) {
			this.tokenPromise = this.bootstrapToken()
				.then((token) => {
					this.currentToken = token;
					return token;
				})
				.finally(() => {
					this.tokenPromise = undefined;
				});
		}
		return this.tokenPromise;
	}

	close(): void {}

	async request(method: string, path: string, options: RequestOptions = {}): Promise<Response> {
		const url = new URL(`${this.baseUrl}/v1${path}`);
		appendParams(url, options.params);
		const headers = await this.authHeaders();
		if (options.headers) {
			for (const [key, value] of Object.entries(options.headers)) {
				headers.set(key, value);
			}
		}

		let body: BodyInit | undefined;
		if (options.content !== undefined) {
			body = options.content;
		} else if (options.jsonBody !== undefined) {
			body = JSON.stringify(options.jsonBody);
			headers.set("Content-Type", "application/json");
		}

		const timeout = options.timeout ?? this.timeout;
		const idempotent = options.idempotent ?? true;
		const readOnly = options.readOnly ?? false;
		for (let attempt = 0; attempt <= this.maxRetries; attempt += 1) {
			let response: Response;
			try {
				response = await this.fetchWithTimeout(url, { method, headers, body }, timeout);
			} catch (error) {
				if (attempt >= this.maxRetries || (!idempotent && !readOnly)) {
					throw new TransportError(`network error after ${attempt + 1} attempts: ${formatError(error)}`);
				}
				await sleepBackoff(attempt);
				continue;
			}

			if (
				retryStatus.has(response.status) &&
				attempt < this.maxRetries &&
				(await shouldRetryResponse(response, {
					idempotent,
					readOnly,
				}))
			) {
				const retryAfter = parseRetryAfter(response);
				if (retryAfter !== undefined) {
					await sleep(Math.min(retryAfter * 1000, 30_000));
				} else {
					await sleepBackoff(attempt);
				}
				continue;
			}

			if (response.status >= 400) {
				await raiseForStatus(response);
			}
			return response;
		}

		throw new TransportError("unexpected retry exhaustion");
	}

	async stream(method: string, path: string, options: Omit<RequestOptions, "content"> = {}): Promise<Response> {
		const url = new URL(`${this.baseUrl}/v1${path}`);
		appendParams(url, options.params);
		const headers = await this.authHeaders();
		if (options.headers) {
			for (const [key, value] of Object.entries(options.headers)) {
				headers.set(key, value);
			}
		}
		let body: BodyInit | undefined;
		if (options.jsonBody !== undefined) {
			body = JSON.stringify(options.jsonBody);
			headers.set("Content-Type", "application/json");
		}
		const response = await this.fetchWithTimeout(url, { method, headers, body }, options.timeout ?? this.timeout);
		if (response.status >= 400) {
			await raiseForStatus(response);
		}
		return response;
	}

	private async bootstrapToken(): Promise<string> {
		const body: Record<string, unknown> = { sub: this.sub, expiresIn: this.expiresIn };
		let url: string;
		const headers = new Headers({ "User-Agent": this.userAgent, "Content-Type": "application/json" });
		if (this.authSecret !== undefined) {
			if (this.tenant) {
				body.tenant = this.tenant;
			}
			url = `${this.baseUrl}/v1/auth/bootstrap`;
			headers.set("X-Auth-Secret", this.authSecret);
		} else {
			url = `${this.baseUrl}/v1/auth/admin`;
			headers.set("X-Admin-Secret", this.adminSecret!);
			headers.set("Authorization", `Bearer ${this.adminBearerToken!}`);
		}

		const response = await this.fetchWithTimeout(
			url,
			{ method: "POST", headers, body: JSON.stringify(body) },
			this.timeout,
		);
		if (response.status !== 201) {
			await raiseForStatus(response);
		}
		const payload = await readJsonObject(response);
		if (typeof payload.token !== "string") {
			throw new AuthError("bootstrap response missing token", { status: response.status, details: payload });
		}
		return payload.token;
	}

	private async authHeaders(): Promise<Headers> {
		return new Headers({
			Authorization: `Bearer ${await this.getToken()}`,
			"User-Agent": this.userAgent,
		});
	}

	private async fetchWithTimeout(input: string | URL, init: RequestInit, timeout: number): Promise<Response> {
		const controller = new AbortController();
		const timer = setTimeout(() => controller.abort(), timeout);
		try {
			return await this.fetchImpl(input, { ...init, signal: controller.signal });
		} finally {
			clearTimeout(timer);
		}
	}
}

interface ParsedError {
	body: Record<string, unknown>;
	code?: string;
	message: string;
	details?: unknown;
}

async function raiseForStatus(response: Response): Promise<never> {
	const parsed = await parseErrorBody(response);
	const options = { code: parsed.code, status: response.status, details: parsed.details };
	if (response.status === 401 || response.status === 403) {
		throw new AuthError(parsed.message, options);
	}
	if (response.status === 404) {
		throw new NotFoundError(parsed.message, options);
	}
	if (response.status === 408) {
		throw new ExecTimeoutError(parsed.message, {
			...options,
			durationMs: typeof parsed.body.durationMs === "number" ? parsed.body.durationMs : undefined,
		});
	}
	if (response.status === 409) {
		throw new ConflictError(parsed.message, options);
	}
	if (response.status === 400 || response.status === 422) {
		throw new ValidationError(parsed.message, options);
	}
	if (response.status === 429) {
		throw new RateLimitError(parsed.message, { ...options, retryAfter: parseRetryAfter(response) });
	}
	if (response.status >= 500 && response.status < 600) {
		throw new ServerError(parsed.message, options);
	}
	throw new SQLFSError(parsed.message, options);
}

async function parseErrorBody(response: Response): Promise<ParsedError> {
	const body = await readJsonObject(response.clone());
	const code = typeof body.code === "string" ? body.code : undefined;
	const message = typeof body.error === "string" ? body.error : `HTTP ${response.status}`;
	return { body, code, message, details: body.details };
}

async function shouldRetryResponse(
	response: Response,
	options: { idempotent: boolean; readOnly: boolean },
): Promise<boolean> {
	if (response.status === 503 && !options.readOnly) {
		const parsed = await parseErrorBody(response.clone());
		if (parsed.code === "ECOHERENCE") {
			return false;
		}
	}
	if (!options.idempotent && !options.readOnly) {
		return false;
	}
	return true;
}

function parseRetryAfter(response: Response): number | undefined {
	const raw = response.headers.get("Retry-After");
	if (raw === null) {
		return undefined;
	}
	const parsed = Number.parseInt(raw, 10);
	return Number.isNaN(parsed) ? undefined : parsed;
}

function appendParams(url: URL, params?: Record<string, unknown>): void {
	if (!params) {
		return;
	}
	for (const [key, value] of Object.entries(params)) {
		if (value !== undefined && value !== null) {
			url.searchParams.set(key, String(value));
		}
	}
}

function sleep(ms: number): Promise<void> {
	return new Promise((resolve) => setTimeout(resolve, ms));
}

async function sleepBackoff(attempt: number): Promise<void> {
	const base = 250;
	await sleep(Math.min(Math.random() * base * 2 ** attempt, 8000));
}

function formatError(error: unknown): string {
	return error instanceof Error ? error.message : String(error);
}

async function* iterLines(response: Response): AsyncGenerator<string> {
	if (!response.body) {
		return;
	}
	const reader = response.body.getReader();
	const decoder = new TextDecoder();
	let buffer = "";

	try {
		while (true) {
			const { value, done } = await reader.read();
			if (done) {
				break;
			}
			buffer += decoder.decode(value, { stream: true });
			let newlineIndex = buffer.search(/\r?\n/);
			while (newlineIndex >= 0) {
				const rawLine = buffer.slice(0, newlineIndex);
				const newlineLength = buffer[newlineIndex] === "\r" && buffer[newlineIndex + 1] === "\n" ? 2 : 1;
				buffer = buffer.slice(newlineIndex + newlineLength);
				yield rawLine.endsWith("\r") ? rawLine.slice(0, -1) : rawLine;
				newlineIndex = buffer.search(/\r?\n/);
			}
		}
		buffer += decoder.decode();
		if (buffer.length > 0) {
			yield buffer.endsWith("\r") ? buffer.slice(0, -1) : buffer;
		}
	} finally {
		reader.releaseLock();
	}
}

function parseSsePayload(dataLines: string[]): unknown {
	const raw = dataLines.join("\n");
	try {
		return JSON.parse(raw);
	} catch {
		return { raw };
	}
}

export async function* iterSseEvents(response: Response): AsyncGenerator<[string, unknown]> {
	let eventName = "message";
	let dataLines: string[] = [];

	for await (const line of iterLines(response)) {
		if (line === "") {
			if (dataLines.length > 0) {
				yield [eventName, parseSsePayload(dataLines)];
				eventName = "message";
				dataLines = [];
			}
		} else if (line.startsWith(":")) {
			// Ignore SSE comments and heartbeat lines.
		} else if (line.startsWith("event:")) {
			eventName = line.slice("event:".length).trim();
		} else if (line.startsWith("data:")) {
			dataLines.push(line.slice("data:".length).replace(/^ /, ""));
		}
	}

	if (dataLines.length > 0) {
		yield [eventName, parseSsePayload(dataLines)];
	}
}
